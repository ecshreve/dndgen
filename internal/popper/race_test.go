package popper_test

import (
	"context"
	"testing"

	"entgo.io/ent/dialect/sql/schema"
	"github.com/ecshreve/dndgen/ent"
	"github.com/ecshreve/dndgen/internal/popper"
	"github.com/ecshreve/dndgen/internal/utils"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"

	_ "github.com/mattn/go-sqlite3"
)

type testRace struct {
	Indx string `json:"index"`
	Name string `json:"name"`
}

func TestRacePopulator(t *testing.T) {
	var data []popper.RaceJSON

	err := utils.LoadJSONFile("testdata/simpleRace.json", &data)
	require.NoError(t, err)

	snaps.MatchJSON(t, data)
}

func TestRacePopulatorFull(t *testing.T) {
	ctx := context.Background()
	// cl, err := ent.Open("sqlite3", "file:ent?cache=shared&mode=memory&_fk=1")
	cl, err := ent.Open("sqlite3", "file:testdb.db?_fk=1")
	require.NoError(t, err)

	err = cl.Schema.Create(ctx, schema.WithGlobalUniqueID(true))
	require.NoError(t, err)

	pp := popper.NewPopper(ctx, cl, "data")
	err = pp.PopulateAll(ctx)
	require.NoError(t, err)

	profPop := popper.NewProficiencyPopulator(pp, "data/Proficiency.json")
	err = profPop.Populate(ctx)
	require.NoError(t, err)

	racePop := popper.NewRacePopulator(pp, "testdata/simpleRace.json")
	err = racePop.Populate(ctx)
	require.NoError(t, err)

	allRaces, err := cl.Race.Query().
		WithStartingProficiencies().
		WithStartingProficiencyOptions(
			func(pc *ent.ProficiencyChoiceQuery) {
				pc.WithProficiencies()
			},
		).
		All(ctx)
	require.NoError(t, err)

	snaps.MatchJSON(t, allRaces)
}
