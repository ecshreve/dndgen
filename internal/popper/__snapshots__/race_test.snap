
[TestRacePopulator - 1]
[
 {
  "age": "Dwarves mature at the same rate as humans, but they're considered young until they reach the age of 50. On average, they live about 350 years.",
  "alignment": "Most dwarves are lawful, believing firmly in the benefits of a well-ordered society. They tend toward good as well, with a strong sense of fair play and a belief that everyone deserves to share in the benefits of a just order.",
  "index": "dwarf",
  "language_desc": "You can speak, read, and write Common and Dwarvish. Dwarvish is full of hard consonants and guttural sounds, and those characteristics spill over into whatever other language a dwarf might speak.",
  "name": "Dwarf",
  "proficiencies": [
   {
    "index": "battleaxes"
   },
   {
    "index": "handaxes"
   },
   {
    "index": "light-hammers"
   },
   {
    "index": "warhammers"
   }
  ],
  "size": "Medium",
  "size_description": "Dwarves stand between 4 and 5 feet tall and average about 150 pounds. Your size is Medium.",
  "speed": 25,
  "starting_proficiency_options": {
   "choose": 1,
   "desc": [
    "You gain proficiency with the artisan’s tools of your choice: smith’s tools, brewer’s supplies, or mason’s tools."
   ],
   "from": {
    "option_set_type": "options_array",
    "options": [
     {
      "item": {
       "index": "smiths-tools"
      },
      "option_type": "reference"
     },
     {
      "item": {
       "index": "brewers-supplies"
      },
      "option_type": "reference"
     },
     {
      "item": {
       "index": "masons-tools"
      },
      "option_type": "reference"
     }
    ]
   },
   "type": "proficiencies"
  }
 }
]
---

[TestRacePopulatorFull - 1]
[
 {
  "ability_bonuses": [
   {
    "ability_score": {
     "desc": [
      "Constitution measures health, stamina, and vital force.",
      "Constitution checks are uncommon, and no skills apply to Constitution checks, because the endurance this ability represents is largely passive rather than involving a specific effort on the part of a character or monster."
     ],
     "full_name": "Constitution",
     "id": 4294967299,
     "index": "con",
     "name": "CON"
    },
    "bonus": 2,
    "id": 1
   }
  ],
  "age": "Dwarves mature at the same rate as humans, but they're considered young until they reach the age of 50. On average, they live about 350 years.",
  "alignment": "Most dwarves are lawful, believing firmly in the benefits of a well-ordered society. They tend toward good as well, with a strong sense of fair play and a belief that everyone deserves to share in the benefits of a just order.",
  "id": 77309411329,
  "index": "dwarf",
  "language_desc": "You can speak, read, and write Common and Dwarvish. Dwarvish is full of hard consonants and guttural sounds, and those characteristics spill over into whatever other language a dwarf might speak.",
  "name": "Dwarf",
  "size": "Medium",
  "size_description": "Dwarves stand between 4 and 5 feet tall and average about 150 pounds. Your size is Medium.",
  "speed": 25,
  "starting_proficiencies": [
   {
    "id": 64424509457,
    "indx": "handaxes",
    "name": "Handaxes",
    "reference": "/api/equipment/handaxe"
   },
   {
    "id": 64424509459,
    "indx": "light-hammers",
    "name": "Light hammers",
    "reference": "/api/equipment/light-hammer"
   },
   {
    "id": 64424509468,
    "indx": "battleaxes",
    "name": "Battleaxes",
    "reference": "/api/equipment/battleaxe"
   },
   {
    "id": 64424509484,
    "indx": "warhammers",
    "name": "Warhammers",
    "reference": "/api/equipment/warhammer"
   }
  ],
  "starting_proficiency_options": {
   "choose": 1,
   "desc": [
    "You gain proficiency with the artisan’s tools of your choice: smith’s tools, brewer’s supplies, or mason’s tools."
   ],
   "id": 68719476737,
   "proficiencies": [
    {
     "id": 64424509492,
     "indx": "brewers-supplies",
     "name": "Brewer's Supplies",
     "reference": "/api/equipment/brewers-supplies"
    },
    {
     "id": 64424509501,
     "indx": "masons-tools",
     "name": "Mason's Tools",
     "reference": "/api/equipment/masons-tools"
    },
    {
     "id": 64424509504,
     "indx": "smiths-tools",
     "name": "Smith's Tools",
     "reference": "/api/equipment/smiths-tools"
    }
   ]
  }
 }
]
---
